<a href="https://colab.research.google.com/github/shriyan44/Lab-6-DATA-301/blob/main/Copy_of_DATA_301_Lab_6B_Shriya_Nimmagadda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Bing Maps API

In this part of the lab, you will join the CityBikes data from the previous part of the lab with additional data that you will query from the Bing Maps API.

First, you will need to register for a Bing Maps Key. Follow the instructions [here](https://docs.microsoft.com/en-us/bingmaps/getting-started/bing-maps-dev-center-help/getting-a-bing-maps-key). You should be able to sign in with your Cal Poly account. When you get to Step 4, select:

- Key type: Basic
- Application type: Dev/Test

You will be able to make 125000 free requests with the resulting API key. This should be more than enough to complete this assignment.

We will be working with the [REST services in the Bing Maps API](https://docs.microsoft.com/en-us/bingmaps/rest-services/). Click on the link for a complete documentation of the features.

## Question 1

Read in the `DataFrame` of bike stations in the United States from Part A of this lab. Restrict to the stations in the "Bay Wheels" network (with network ID "ford-gobike").

How many of these stations are in the city/county of San Francisco? 

(_Hint:_ Use the [Locations API](https://docs.microsoft.com/en-us/bingmaps/rest-services/locations/) to get the address associated with each latitude and longitude coordinate.)

In [1]:

# AjDIyFMNcKFfc6oWn8OmF_gxlT8rUMkepcqE6YB2z2EsGyyRFQMSAMXVoKNFH8Lq
from google.colab import drive
import pandas as pd
drive.mount('/content/drive')
df_stations = pd.read_csv("/content/drive/My Drive/stations.csv")
df_bay = df_stations[df_stations["networkid"] == "bay-wheels"]
# df_bay.reset_index(drop=True)
df_bay


Mounted at /content/drive


Unnamed: 0  empty_slots  free_bikes  ... extra.slots  networkname   networkid
6230        6230         11.0           4  ...         NaN   Bay Wheels  bay-wheels
6231        6231          3.0           8  ...         NaN   Bay Wheels  bay-wheels
6232        6232          3.0          20  ...         NaN   Bay Wheels  bay-wheels
6233        6233          1.0          14  ...         NaN   Bay Wheels  bay-wheels
6234        6234         22.0           4  ...         NaN   Bay Wheels  bay-wheels
...          ...          ...         ...  ...         ...          ...         ...
6707        6707         18.0           5  ...         NaN   Bay Wheels  bay-wheels
6708        6708         17.0          10  ...         NaN   Bay Wheels  bay-wheels
6709        6709          9.0           6  ...         NaN   Bay Wheels  bay-wheels
6710        6710          6.0          13  ...         NaN   Bay Wheels  bay-wheels
6711        6711         14.0           9  ...         NaN   Bay Wheels  bay-wheels

[482 rows x 20 columns]

In [2]:
df_bay1 = df_bay.loc[[6232,6234]]
df_bay1

Unnamed: 0  empty_slots  free_bikes  ... extra.slots  networkname   networkid
6232        6232          3.0          20  ...         NaN   Bay Wheels  bay-wheels
6234        6234         22.0           4  ...         NaN   Bay Wheels  bay-wheels

[2 rows x 20 columns]

In [3]:
import requests
import time
import pandas as pd
locinfo = []

for i in df_bay.index:
  response = requests.get("http://dev.virtualearth.net/REST/v1/Locations/" + str(df_bay["latitude"][i]) 
  + "," + str(df_bay["longitude"][i]) + "?includeEntityTypes=address&key=AjDIyFMNcKFfc6oWn8OmF_gxlT8rUMkepcqE6YB2z2EsGyyRFQMSAMXVoKNFH8Lq")
  locinfo.append(response.json()["resourceSets"][0])
  # print(locinfo)
  # print(i)
  time.sleep(0.2)



# df_sf = df_all[df_all["San Francisco"]]
# df_sf





In [9]:
df_all = pd.json_normalize(locinfo, "resources", errors="ignore")
df_all["stationName"] = df_bay["name"].to_numpy()
df_all["latitude"] = df_bay["latitude"].to_numpy()
df_all["longitude"] = df_bay["longitude"].to_numpy()
df_all.head()


__type  ...   longitude
0  Location:http://schemas.microsoft.com/search/l...  ... -122.270582
1  Location:http://schemas.microsoft.com/search/l...  ... -121.889277
2  Location:http://schemas.microsoft.com/search/l...  ... -121.881090
3  Location:http://schemas.microsoft.com/search/l...  ... -121.874119
4  Location:http://schemas.microsoft.com/search/l...  ... -121.877120

[5 rows x 24 columns]

In [10]:
df_sf = df_all[df_all["address.adminDistrict2"] == "San Francisco Co."]
df_sf


__type  ...   longitude
25   Location:http://schemas.microsoft.com/search/l...  ... -122.426497
26   Location:http://schemas.microsoft.com/search/l...  ... -122.390833
27   Location:http://schemas.microsoft.com/search/l...  ... -122.436861
28   Location:http://schemas.microsoft.com/search/l...  ... -122.398870
30   Location:http://schemas.microsoft.com/search/l...  ... -122.451122
..                                                 ...  ...         ...
477  Location:http://schemas.microsoft.com/search/l...  ... -122.443238
478  Location:http://schemas.microsoft.com/search/l...  ... -122.479462
479  Location:http://schemas.microsoft.com/search/l...  ... -122.476561
480  Location:http://schemas.microsoft.com/search/l...  ... -122.455102
481  Location:http://schemas.microsoft.com/search/l...  ... -122.435474

[271 rows x 24 columns]

There are 271 SF stations out of 482 stations in the bay-wheels network in US

## Question 2

You want to go to Coit Tower. To save money, you decide to ride a "Bay Wheels" bike to the closest station and hail a cab from there to Coit Tower. What station should you bike to so that you are as close to Coit Tower as possible (as measured by driving distance)? Does your answer agree with the one that you obtained in Part A of this lab? If not, why does it differ?

_Hints:_ 
- You should restrict your attention to bike stations that are in San Francisco, which you determined in Question 1. 
- Use the [Routes API](https://docs.microsoft.com/en-us/bingmaps/rest-services/routes/calculate-a-distance-matrix) to calculate a distance matrix between Coit Tower and the bike stations. 
- You can do this with just one call to the API. Because there are too many stations, it is impossible to specify all the locations in the URL. Instead, you should make a POST request (`requests.post`), passing in the parameters as a JSON object through the `json=` parameter of `requests.post`. Read the API documentation carefully to learn how to use the POST API.

In [11]:
# 37.802747 and longitude -122.405861
def build_request_body():
  request_body = {
      "origins": [],
      "destinations": [{
      "latitude": 37.8024,
      "longitude": -122.4058
      }],
      "travelMode": "Driving"
  }
  for row_series in df_sf.iterrows():
    request_body["origins"].append({
        "latitude": row_series[1]["latitude"],
        "longitude": row_series[1]["longitude"]
        })

  return request_body

response = requests.post("https://dev.virtualearth.net/REST/v1/Routes/DistanceMatrix?key=AjDIyFMNcKFfc6oWn8OmF_gxlT8rUMkepcqE6YB2z2EsGyyRFQMSAMXVoKNFH8Lq", json=build_request_body())
df_distance_bike_share = pd.json_normalize(response.json()["resourceSets"][0]["resources"][0]["results"])
df_distance_bike_share = df_distance_bike_share.join(df_sf.reset_index())

min_index = df_distance_bike_share["travelDistance"].idxmin()
df_distance_bike_share.loc[min_index]

destinationIndex                                                                         0
originIndex                                                                            218
totalWalkDuration                                                                        0
travelDistance                                                                       0.965
travelDuration                                                                      3.2333
index                                                                                  416
__type                                   Location:http://schemas.microsoft.com/search/l...
bbox                                     [37.798896982429326, -122.41975667826308, 37.8...
name                                     710 Lombard St, San Francisco, CA 94133, Unite...
confidence                                                                            High
entityType                                                                         Address

The closest staion is on Lombard St. at Columbus Avenue. I got a different answer than from lab A because this route api takes the actualy driving route into account so it's not just based on manhattan distance so we get a more realistic answer.

## Submission Instructions

- Copy this notebook to your own Drive, if you have not already.
- Restart this notebook and run the cells from beginning to end. 
  - Go to Runtime > Restart and Run All.
- Rename this notebook by clicking on "DATA 301 Lab 6B - YOUR NAMES HERE" at the very top of this page. Replace "YOUR NAMES HERE" with the first and last names of you (and your partners, for Phase 2).
- Get the link to your notebook:
  - Click on "Share" at the top-right. 
  - Change the settings to "Anyone with the link can view". 
  - Copy the sharing link into Canvas.